### ISSUE TITLE
Merge apicaffe_cleanup branch into develop

### ISSUE THREAD

**Issue by drinkingkazu** [LArbys/LArCV#43]
The branch apicaffe_cleanup has a lot of enhancements (segmentation training and full implementation of ProcessDriver for caffe training) while it is supposed to support old features as well.  We have to merge back into master at some point.

We must make sure to test with:

*) Supera
*) Neutrino + Cosmic overlay image generation
*) Semantic Segmentation training
*) Single particle classification training

**Comment by vgenty** [comment]
I have a motivation to test out the last bullet

> *) Single particle classification

en route to training DeconvNet so i'll do that

**Comment by twongjirad** [comment]
I will test supera on this new branch, as I have just successfully made images checked in the viewer.

**Comment by twongjirad** [comment]
Was planning to test Supera on the GPVM where current data larlite files reside. 

Getting this compilation error. Looking into it. But maybe one of you knows where to look.

[ tmw@uboonegpvm03 Base ]$ make
make: **\* No rule to make target `install_header', needed by`main_build'.  Stop.

in GNUMakefile.CORE

problem comes from this:

```
install_header: $(OBJECTS) $(DICT).o
        @echo '<< installing headers >>'
        @$(foreach f, $(HEADERS),$(shell cp $(f) $(PACKAGE_INCDIR)/))
```

which is a rule that is only defined if ROOT6 exists, but not if ROOT5 exists. 

@drinkingkazu  says he knows what to do

**Comment by twongjirad** [comment]
According to @drinkingkazu this branch really should only be applied to ROOT6.  ROOT5 will take some work.

**Comment by drinkingkazu** [comment]
@twongjirad if you can test Supera w/ ROOT6 that would be actually great. I can work on ROOT5 support in the mean time unless you started already.

**Comment by drinkingkazu** [comment]
merged #44

**Comment by vgenty** [comment]
Single particle training seems to work OK. Trained about half an epoch for alex 5 particle high resolution.
- Usual alex with dropout, no batch norm, from scratch, rms prop

![train_log_0](https://cloud.githubusercontent.com/assets/5595000/16996695/65c7c89e-4e80-11e6-98f5-da30529bd3b3.png)
- Alex with no dropout, batch norm after each conv, fine tune model from public note, rms prop  

![train_log_1](https://cloud.githubusercontent.com/assets/5595000/16996700/69a3c698-4e80-11e6-86d8-64cf8c2e4fd6.png)
- Alex with no dropout, batch norm after each conv, fine tune model from public note, SGD  

![train_log_2](https://cloud.githubusercontent.com/assets/5595000/16996701/69a4ad9c-4e80-11e6-9d13-acbcf86bab40.png)
- Alex with no dropout, batch norm after each conv, from scratch, SGD  

![train_log_3](https://cloud.githubusercontent.com/assets/5595000/16996699/699e085c-4e80-11e6-86be-50aecdfab2f1.png)

**Comment by drinkingkazu** [comment]
OK ROOT5 build is now supported. Checked ROOT6 still builds just fine. Alongside fixed a bug to generate .depend: people has to remove .depend individually unfortunately but then one can benefit from it (avoids re-compilation of some packages under app).

**Comment by drinkingkazu** [comment]
pull request into develop made #55

**Comment by drinkingkazu** [comment]
Actually there will be a backward-incompatible change in caffe configuration file w/ apicaffe_cleanup branch. Might be a good idea to make a tag from current develop.
